core/src/main/java/com/linecorp/armeria/server/AnnotatedValueResolver.java

```diff
@@ -0,0 +1,764 @@
+                                               List<RequestObjectResolver> objectResolvers) {
+        return of(parameter, parameter, parameter.getType(), pathParams, objectResolvers);
+    }
+
+    /**
+     * Returns a list of {@link AnnotatedValueResolver} which is constructed with the specified
+     * {@link Field}, {@code pathParams} and {@code objectResolvers}.
+     */
+    static Optional<AnnotatedValueResolver> of(Field field, Set<String> pathParams,
+                                               List<RequestObjectResolver> objectResolvers) {
+        return of(field, field, field.getType(), pathParams, objectResolvers);
+    }
+
+    /**
+     * Creates a new {@link AnnotatedValueResolver} instance if the specified {@code annotatedElement} is
+     * a component of {@link AnnotatedHttpService}.
+     *
+     * @param annotatedElement an element which is annotated with a value specifier such as {@link Param} and
+     *                         {@link Header}.
+     * @param typeElement      an element which is used for retrieving its type and name.
+     * @param type             a type of the given {@link Parameter} or {@link Field}. It is a type of
+     *                         the specified {@code typeElement} parameter.
+     * @param pathParams       a set of path variables.
+     * @param objectResolvers  a list of {@link RequestObjectResolver} to be evaluated for the objects which
+     *                         are annotated with {@link RequestObject} annotation.
+     */
+    private static Optional<AnnotatedValueResolver> of(AnnotatedElement annotatedElement,
+                                                       AnnotatedElement typeElement, Class<?> type,
+                                                       Set<String> pathParams,
+                                                       List<RequestObjectResolver> objectResolvers) {
+        requireNonNull(annotatedElement, "annotatedElement");
+        requireNonNull(typeElement, "typeElement");
+        requireNonNull(type, "type");
+        requireNonNull(pathParams, "pathParams");
+        requireNonNull(objectResolvers, "objectResolvers");
+
+        final Param param = annotatedElement.getAnnotation(Param.class);
+        if (param != null) {
+            final String name = findName(param, typeElement);
+            if (pathParams.contains(name)) {
+                final AnnotatedValueResolver element =
+                        ofHttpComponent(param, name, typeElement, type,
+                                        annotatedElement.getAnnotation(Default.class),
+                                        httpPathOrParamValueResolver(ctx -> ctx.context().pathParam(name)),
+                                        false, AggregationStrategy.NONE);
+                if (!element.isRequired()) {
+                    throw new IllegalArgumentException(
+                            "Path variable '" + name + "' should not be an optional.");
```

Comment:
- `Path variable 'foo' should not be an optional.` -> `Path variable 'foo' annotated with @Default` cannot be an Optional.`
- Using `@Default` on an `Optional` parameter is not optimal, but it's probably OK to support such a use case? We may want to log some warning though.